core/node/state_keeper/src/updates/mod.rs

```diff
@@ -77,6 +79,20 @@ impl UpdatesManager {
         self.base_system_contract_hashes
     }
 
+    pub(crate) fn get_next_l2_block_params_or_batch_params(&mut self) -> L2BlockParams {
```

Comment:
AFAICT, this method should be replaced by providing a mutable ref to the timestamp from the next block params. It must not have a fallback; the caller is responsible for calling this method only when the next block params exist (as the state keeper would).

Reply:
You mean the first transaction in a batch, right? AFAIU, the current implementation won't set next L2 block params on `UpdatesManager` initialization (= when opening a new batch), meaning that the timestamp of the first block in a batch may still be outdated. Is this the case? If so, it can be solved incrementally, but I wouldn't consider a fix complete without covering this scenario. I *think* there's no choice here other than delaying creating `UpdatesManager` / `BatchExecutor` until we have a first transaction to execute (because the first block in the batch is immediately initialized in the VM and it cannot be modified afterwards), and this would require some refactoring of the state keeper logic.